### DISCUSSION TITLE
Optimize search feature

### DISCUSSION THREAD

**Discussion by thang-qt** [imfing/hextra#112] [General]
Hi.
I'm trying to build a knowledge sharing site. I'm testing with about 4000 pages and my machine would crash instantly if I try to search even it's a really fast one. Is there anyway to optimize search speed or config alternative search engine?

**Comment by remysheppard** [comment]
Are you saying that you're running Hextra, building out 4,000 pages, trying to search, and it **is in fact crashing**? 

Or are you saying you _assume_ it would crash?

Also, an application of that size, you might actually be better off with a database solution than a static site.

**Comment by thang-qt** [comment]
> Are you saying that you're running Hextra, building out 4,000 pages, trying to search, and it **is in fact crashing**?
> 
> Or are you saying you _assume_ it would crash?
> 
> Also, an application of that size, you might actually be better off with a database solution than a static site.

Now that I try again, it in fact not crash but really slow on my laptop. And on my phone, it just freeze the tab.

**Comment by thang-qt** [comment]
And it's only slow on first search once the page load.

**Comment by thang-qt** [comment]
It seems to load faster on chromium browser.
If you want, I can provide you the dev site link in private

**Comment by imfing** [comment]
It would be nice if you could share your repository in public, so I could figure out what's really going on.

I think 4000 pages definitely gonna overload the local search feature in Hextra.
Hextra pre-computes the search index file and when you open the search, the website load it on-the-fly. 4000 pages would result in very large index file including **all the contents**, thus the browsers or the search library just couldn't handle it. 

In this case, I think integrating with third-party cloud offerings such as [Algolia](https://www.algolia.com/) or [Google Custom Search](https://programmablesearchengine.google.com/) might be a better choice.

**Comment by thang-qt** [comment]
> It would be nice if you could share your repository in public, so I could figure out what's really going on.
> 
> I think 4000 pages definitely gonna overload the local search feature in Hextra. Hextra pre-computes the search index file and when you open the search, the website load it on-the-fly. 4000 pages would result in very large index file including **all the contents**, thus the browsers or the search library just couldn't handle it.
> 
> In this case, I think integrating with third-party cloud offerings such as [Algolia](https://www.algolia.com/) or [Google Custom Search](https://programmablesearchengine.google.com/) might be a better choice.

I'm not gonna expose the repo in public.
But if you're up to implementing a alternative search engine. I might suggest [Pagefind](https://pagefind.app/) as I tried it locally and it's definitely much faster than the current one

**Reply by imfing** [reply]
Didn't know about [Pagefind](https://pagefind.app/) but I'll try it out.
I wonder if that scales well to 4000 pages with full-text contents?

> I tried it locally and it's definitely much faster than the current one

how fast? it doesn't compile the site pages, so it's expected.

**Reply by thang-qt** [reply]
Sorry that I missed this reply. The results appear almost instantly as I type so I'd say that It's pretty fast. The downsize is that it needs the site to be compiled to build the search index. And the building index process took even longer than the time it takes to build the site. So faster for users but slower for maintainer.
Another suggestion is that you add an option so that user can search title only, it'd be way faster.

**Reply by thang-qt** [reply]
Not sure if this is the right way but I modify search-data.json as follow ($data to $pageTitle)
```
{{- $output = $output | merge (dict $pageLink (dict "title" $pageTitle "data" $pageTitle)) -}}
```
It did bring the final json from 15MB to less than 1MB. However, the initial click on search box still slow down the browser and on the phone, it freeze the tab for around 5s. I can see that it faster than indexing full data, though

**Reply by imfing** [reply]
It makes sense to only include titles when there are thousands pages. I can provide a flag for this in the future.

> Not sure if this is the right way but I modify search-data.json as follow ($data to $pageTitle)

I doubt this would work correctly as we may also need to modify the search js script.

**Reply by thang-qt** [reply]
Do you have plan to implement title only search in near future? I just tested several theme using flexsearch and they're also laggy

**Reply by imfing** [reply]
> Do you have plan to implement title only search in near future? I just tested several theme using flexsearch and they're also laggy

By "laggy", do you mean using title only search or using full text search?

**Reply by imfing** [reply]
@thangisme Can you try the following to include only the title without the page data:

```diff
diff --git a/assets/json/search-data.json b/assets/json/search-data.json
index 16d3ad7..229e668 100644
--- a/assets/json/search-data.json
+++ b/assets/json/search-data.json
@@ -7,7 +7,7 @@
 {{- range $index, $page := $pages -}}
   {{- $pageTitle := $page.LinkTitle | default $page.File.BaseFileName -}}
   {{- $pageLink := $page.RelPermalink -}}
-  {{- $data := partial "utils/fragments" $page -}}
+  {{- $data := dict "" "" -}}
   {{- $output = $output | merge (dict $pageLink (dict "title" $pageTitle "data" $data)) -}}
 {{- end -}}
```

just a one-line change to override the `assets/json/search-data.json`  as shown above, which should produce a much smaller search index json file.

**Reply by thang-qt** [reply]
That did it. Thaank you!

**Reply by xuanswe** [reply]
> Sorry that I missed this reply. The results appear almost instantly as I type so I'd say that It's pretty fast. The downsize is that it needs the site to be compiled to build the search index. And the building index process took even longer than the time it takes to build the site. So faster for users but slower for maintainer.

Hi @thangisme, @imfing,

Is it possible to configure Hextra to use another search engine, that builds the indexes at compile time, like the mentioned [Pagefind](https://pagefind.app)? If yes, how could we configure with Hextra? Thanks!

If everything works, then I am thinking if we can have a CI/CD like below:
* First, we deploy only hugo site without building search indexes.
* Then we have another job to build search indexes in parallel and push the result later to the hugo site.
This will allow the building site fast, while the search feature will be available later.

**Reply by imfing** [reply]
Hi @xuanswe 
Pagefind looks pretty nice. Guess we can figure out a plan to support that. Can you create a feature request issue?

**Reply by xuanswe** [reply]
Hi @imfing, it's here https://github.com/imfing/hextra/issues/636.

**Comment by imfing** [comment]
FYI, I added configuration options different index types: `content | summary | heading | title` in https://github.com/imfing/hextra/pull/145
